### PR TITLE
v1.2.2

### PR DESCRIPTION
**Pull request by efjacobson**
* Implement `NcrVideoProjector` handlers for transcoding/transcription events.
* Add command handlers for `UpdateTranscodingStatus` and `UpdateTranscriptionStatus`.
* On video update, set `mezzanine_url` and `kaltura_mp4_url` when `mezzanine_ref` has changed.
* On video update, add `en` to `caption_urls` map when `caption_ref` has changed.
* `SyncMediaHandler` now uses `created_at` for start date (instead of `order_date`).

### CODE REVIEW

File: src/Ovp/UpdateTranscriptionStatusHandler.php [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+                ->addToMap('tags', 'document_asset_ref', $documentRef->toString());
+            $streamId = StreamId::fromString(sprintf('%s.history:%s', $linkedRef->getLabel(), $linkedRef->getId()));
+            $pbjx->getEventStore()->putEvents($streamId, [$videoEvent]);
+        }
+    }
+
+    protected function isNodeRefSupported(NodeRef $nodeRef): bool
+    {
+        static $videoAssetLabel = null;
+        if (null === $videoAssetLabel) {
```
Comment (gdbrown) [inline src/Ovp/UpdateTranscriptionStatusHandler.php:128]: @efjacobson just use video-asset check against label

File: src/Ovp/UpdateTranscodingStatusHandler.php [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+            ));
+        }
+        return $searchAssetsRequestClass::create();
+    }
+
+    protected function isNodeRefSupported(NodeRef $nodeRef): bool
+    {
+        static $videoAssetLabel = null;
+        if (null === $videoAssetLabel) {
+            $videoAssetLabel = VideoAssetV1Mixin::create()->getId()->getCurie()->getMessage();
```
[the remark this answers is not in the document]
Comment (efjacobson) [reply src/Ovp/UpdateTranscodingStatusHandler.php]: done

File: src/Ovp/UpdateTranscodingStatusHandler.php [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+        if ($transcodingStatus !== TranscodingStatus::COMPLETED()) {
+            return; // we only update the video if transcoding completed
+        }
+
+        $request = $this->getSearchAssetsRequest()
+            ->set('linked_ref', $videoAssetRef)
+            ->addToSet('types', ['image-asset']);
+        $response = $pbjx->request($request);
+        $images = $response->get('nodes', []);
```
[the remark this answers is not in the document]
Comment (efjacobson) [reply src/Ovp/UpdateTranscodingStatusHandler.php]: done

File: src/Ovp/UpdateTranscodingStatusHandler.php [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+        $videoAsset = $this->ncr->getNode($videoAssetRef);
+        foreach ($videoAsset->get('linked_refs', []) as $linkedRef) {
+            if ('video' !== $linkedRef->getLabel()) {
+                continue;
+            }
+            $videoEvent = clone $event;
+            $videoEvent
+                ->set('node_ref', $linkedRef)
```
[the remark this answers is not in the document]
Comment (efjacobson) [reply src/Ovp/UpdateTranscodingStatusHandler.php:103]: done

File: src/Ovp/UpdateTranscriptionStatusHandler.php [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        $response = $pbjx->request($request);
+        $documents = $response->get('nodes', []);
+
+        if (count($documents) < 1) {
+            return;
+        }
+
+        $documentRef = NodeRef::fromNode($documents[0]);
+        $documentEvent = clone $event;
+        $documentEvent->set('node_ref', $documentRef);
```
[the remark this answers is not in the document]
Comment (efjacobson) [reply src/Ovp/UpdateTranscriptionStatusHandler.php]: done for all

File: src/Ovp/UpdateTranscriptionStatusHandler.php (hunk 1 above)
Comment (efjacobson) [reply src/Ovp/UpdateTranscriptionStatusHandler.php]: done

**Comment by efjacobson** [comment]
@gdbrown what do you think of making an artifactUrlProvider twig function?

**Comment by gdbrown** [comment]
@efjacobson that's fine, like ovp_url or ovp_artifact_url or something?  go for it.

**Comment by efjacobson** [comment]
@gdbrown ok, working on that i realized that the artifactUrlProvider can provide urls for audio (`.wav`) and transcription (`-transcribed.json`), neither of which currently have a public acl. wonder if we should either make them public or remove from provider

**Comment by gdbrown** [comment]
@efjacobson we should change them to public so we can reference them.

**Comment by efjacobson** [comment]
> @efjacobson we should change them to public so we can reference them.

@gdbrown then in that case i think the extension is good, ill change the acl in ovp

**Comment by gdbrown** [comment]
@efjacobson make sure any pbjx->request(...) are using copyContext from the command that is kicking off that request.  this is needed to ensure a permission check is not done and for tracking those requests.  e.g. in getLatestImage

**Comment by gdbrown** [comment]
@efjacobson now that twig is directly used in this lib, add it to composer json `"twig/twig": "^2.7"`

**Comment by gdbrown** [comment]
@efjacobson typehint both providers because otherwise overriding at site level won't work.

```
    private ArtifactUrlProvider $artifactUrlProvider;

    public function __construct(DamUrlProvider $damUrlProvider)
    {
        $this->artifactUrlProvider = new ArtifactUrlProvider($damUrlProvider);
    }
```

**Comment by efjacobson** [comment]
> typehint both providers because otherwise overriding at site level won't work.

@gdbrown not sure i follow this...arent they already typehinted? unless you mean put them both in constructor but thats a bit weird too - if artifactUrlProvider is in the constructor then damUrlProvider doesnt need to be

**Comment by gdbrown** [comment]
@efjacobson constructor args with typehints.  avoid new Blah since you can't override that

**Comment by efjacobson** [comment]
@gdbrown got it, done

**Pull request merged**